PR: Get and Cache List of Roles + Helper Functions: has_role?/2

nelsonic opened the pull request:
This Pull Request adds:
+ [x] functions to check if a person has a role: `RBAC.has_role?/2` (`String`) and `RBAC.has_role_any?/2` (`List`)
+ [x] Code to fetch the list of roles from Auth App and cache them in ETS: https://github.com/dwyl/auth/issues/110

nelsonic commented:
@th0mas this PR is in a reviewable state. 🔍  🙏 
It's reasonably independent from https://github.com/dwyl/auth/pull/85 as it _already_ works. ✅ 
Please take a look at the code and if you want to improve any of it, go for it! 👍 
Thanks. ☀️

th0mas commented on README.md:
```diff
@@ -52,20 +54,143 @@ Install by adding `rbac` to your list of dependencies in `mix.exs`:
+# initialize RBAC Cache:
+RBAC.init_roles_cache(
+  "https://dwylauth.herokuapp.com",
+  AuthPlug.Token.client_id()
```
This makes `rbac` implicitly dependent on `auth_plug` - do we want to do this?

nelsonic replied:
Yeah, I thought about this and yes, I don't see _us_ using the modules independently. 💭 
Ultimately both `auth_plug` and `rbac` coupled with `auth` for better or worse.
I just like having separate modules/repos for maintenance purposes. (_they can be independently extended/maintained_)
If someone using any of our modules asks to use them independently, we can always refactor. 

I briefly thought of putting the `init_roles_cache` function in the `auth_plug` ... would that be better? 🤷‍♂️ 

You _have_ reminded me that I haven't explicitly mentioned getting an `AUTH_API_KEY`  in the `rbac/README.md` ... 🤦

th0mas replied:
Putting `init_roles_cache` in auth_plug may be better as then all the functions that work directly with the `auth` server are contained in one place.

However, this also works as it means you can just ignore `rbac` if you want a simple authentication solution, which works for ~80% of people.

Having though about this some more, from a DevX point of view, as well as a "seperation of concerns" point of view, I think its okay to have the dependency on `auth_plug` - but it should be made clear that this *extends* `auth plug`.

nelsonic replied:
@th0mas technically the person doesn't need to depend on `auth_plug` at all.
They can just do the following:

```
apikey = System.get_env("AUTH_API_KEY")
RBAC.init_roles_cache(
  "https://dwylauth.herokuapp.com",
  List.first(String.split(apikey, "/"))
)
```

I just _assume_ that they are using `rbac` with `auth_plug` because that's _our_ current use-case. 💭

nelsonic replied:
also worth noting that in `Elixir 1.11` (_shipping soon_: https://www.youtube.com/watch?v=guc16D_0Imk ...) this type of dependency will need to be explicit. We can no longer _assume_ peer dependencies. So maybe we should update the docs/instructions to this exact code. That way someone can use RBAC 100% independently of `auth_plug`. 👍

nelsonic replied:
I've created a function `insert_roles_into_ets/1` that can be called with a list of roles and added docs/instructions to the `README.md` for how to use it. This makes `RBAC` 100% independent from both `auth` and `auth_plug`. 
It also means that anyone using `auth_plug` doesn't need to use `rbac`. Everything is independent but complimentary. 👍

nelsonic replied:
I've created a function and updated the docs so that `auth_plug` is no longer assumed/required
and people can use `rbac` 100% independently of `auth`:
```elixir
roles = [%{id: 1, name: "admin"}, %{id: 2, name: "subscriber"}]
RBAC.insert_roles_into_ets_cache(roles)
```

th0mas commented on lib/rbac.ex:
```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
+  """
+  def get_approles(auth_url, client_id) do
+    url = "#{auth_url}/approles/#{client_id}"
+    HTTPoison.start()
```
`mix` should start HTTPoison for us, we shouldn't have to call `HTTPoison.start/1` explicitly

nelsonic replied:
Yeah, agreed. For some reason it wasn't working when I was trying it the other day. 
If you have time, please rework it. (_you will need an `AUTH_API_KEY` to run it on `localhost`_)

nelsonic replied:
@th0mas Thanks for making this change. 👍

th0mas commented on lib/rbac.ex:
```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
+  so your app can use the resulting JSON (list of roles).
+  """
+  @spec parse_body_response({atom, String.t()}) :: String.t()
+  def parse_body_response({:error, err}), do: {:error, err}
```
Should these be private functions defined with `defp`? I don't think they will need to be called by any external modules

nelsonic replied:
@yeah, not much danger of them being called externally, but you're 100% correct they should be `defp` as we aren't unit testing them directly. 👍

nelsonic replied:
updated to `defp` ✅

th0mas commented on lib/rbac.ex:
```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
+  def get_role_from_cache(term) do
+    case :ets.lookup(:roles_cache, term) do
+      # not found:
+      [] -> %{id: 0}
```
I think returning an explicit error here instead of a zero value will be clearer

nelsonic replied:
Yeah, this is a **_silent_ fail** when the role is not found. 🤭 
It's probably preferable to give the developer _some_ indication that they typoed the role name. 👍

nelsonic replied:
I've added a `Logger.error` with the stack trace so devs will be informed of a typo in role. 📣
e.g: 
```
09:44:39.356 [error] rbac.ex:112 Role not found in ets cache: admni
```
The way the cache miss does not fail _silently_ so the DevX doesn't suck. 
Very happy for this to be refactored to return an `:error` when the role is not found, 
and for that to be handled by the calling function. 👍 
But simply updating the code to `:error` (_and breaking the tests_) 
without handling the `:error` is counter-productive. 💭

nelsonic replied:
Please open an issue to refactor this if you feel that way inclined. 
Hopefully the `has_role?/2` function is already useful in completing your mission. 👍

th0mas commented on lib/rbac.ex:
```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
+    role = get_role_from_cache(role_name)
+
+    person_roles =
+      conn.assigns.person.roles
```
I personally don't like passing `conn` here.

+ It restricts you to only checking for the currently logged in user.
+ It restricts you to using `auth_plug`
+ You can't check roles of any user you want
+ You have to use this with Phoenix/Plug frameworks

nelsonic replied:
@th0mas yeah, this is a shorthand for our _current_ use-case. 
totally agree that it's not as versatile (_or as simple_) as passing the two pieces of data to be checked.
Please feel free to refactor to accept just the roles (`String` of comma-separated `ints` e.g: `"1,7,42"`) and role `"admin"`

I just thought for our _current_ use-case it would be less verbose than having to type the following:
```elixir
RBAC.has_role?(conn.assigns.person.roles, "admin") 
```
each time ... 💭 

I feel that the _context_ where this will be used is 99% "controllers" in Phoenix.
Unless you think we can use it in the Smart Home Firmware? 🤔

th0mas replied:
It will mostly be used in Phoenix, when identifying users through an NFC tag we look them up in the database separately, which means using `conn` won't work there. It also won't work when using phoenix channels.

nelsonic replied:
Ok. cool. we need to keep our Channels options open so we may want to have a pattern match in the function so that it can accept the `conn` struct as well as a string. 💭

nelsonic replied:
Best of both worlds:
https://github.com/dwyl/rbac/blob/1ea682cc39be950a0c54ed6f2eef3d59da603807/lib/rbac.ex#L134-L142

th0mas commented on lib/rbac.ex:
```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
+    end)
+
+    # list of integers
+    person_roles =
```
DRY - Getting `people_roles` should be moved into a separate function

nelsonic replied:
Yeah, agreed. Can easily be moved out. I can do this now (_unless you are already doing some refactoring ..._) 💭

nelsonic replied:
Created DRY function: 
https://github.com/dwyl/rbac/blob/1ea682cc39be950a0c54ed6f2eef3d59da603807/lib/rbac.ex#L119-L125

th0mas commented:
@nelsonic can you give me write access to this branch so I can push some of my refactoring commits?

nelsonic commented:
@th0mas please confirm you have write access. Thanks. 🔑 
![image](https://user-images.githubusercontent.com/194400/93212633-bbe9cb00-f75a-11ea-9af8-8d9b2ce69f31.png)

th0mas commented:
@nelsonic got it cheers

nelsonic commented:
@th0mas I have made a few updates addressing all your PR review comments. 
I think the code is now more versatile and there's a clear reason for having a _separate_ re-useable module/repo.

`RBAC` no longer makes _any_ assumption of using `auth` or `auth_plug`, it can be used 100% independently/offline.

You can call `RBAC.has_role?([1,14,31], "admin")` with a `List` of `Integers` as the first argument, 
but I have left the `RBAC.has_role?(conn, "admin")` in as a convenience, with the appropriate transformation behind the scenes. Which means the `Channels` use case is still available (_not depending on `Plug.Conn`_).
I want to write `RBAC.has_role_any?(conn, [:superadmin, :admin])` as I feel this code is brief but descriptive. 🤔 

Please take another look at the PR when you're back at your desk tomorrow. 🙏 

**Note**: `get_role_from_cache/1` now logs an error with a stack trace so devs can fix their role name typos. 📣 
`get_role_from_cache/1` is only used internally but the function is `public` in case anyone wants to use it. 
If you prefer to return an explicit `:error` in `get_role_from_cache/1` we need to handle it in the calling function. 💭 

LMK your thoughts. 👍

th0mas commented:
Looks good 👍
